virtool/users/sessions.py

```diff
@@ -2,219 +2,220 @@
-
-    new_session = {
-        "created_at": virtool.utils.timestamp().timestamp(),
-        "ip": ip,
-    }
-
-    token = None
-
-    if user_id:
-        token, hashed = virtool.utils.generate_key()
-        user_document = await db.users.find_one(user_id, session=session)
-        new_session.update(
-            {
-                "token": hashed,
-                "administrator": user_document["administrator"],
-                "groups": user_document["groups"],
-                "permissions": user_document["permissions"],
-                "force_reset": user_document["force_reset"],
-                "user": {"id": user_id},
-            }
-        )
-
-    await redis.set(session_id, dumps(new_session))
-    await redis.expireat(session_id, expires_at.timestamp())
-
-    return session_id, new_session, token
-
-
-async def create_session_id(redis: Redis) -> str:
-    """
-    Create a new unique session id.
+from virtool.api.custom_json import dumps, fromisoformat, isoformat
+from virtool.data.errors import ResourceError
+from virtool.data.piece import DataLayerPiece
+
+
+class SessionData(DataLayerPiece):
+    def __init__(self, redis: Redis):
+        self.redis = redis
+
+    async def create(
+        self,
+        ip: str,
+        user_id: Optional[str] = None,
+        remember: Optional[bool] = False,
+    ) -> Tuple[str, Session, str]:
+        """
+        Creates a new session with the given ip and user_id
+
+        :param ip: the ip address of the client
+        :param user_id: the user id of the client
+        :param remember: Boolean indicating if a session should be saved long term
+        :return: the session id, the session model, and the session token
+        """
+        session_id = await self.create_session_id()
+
+        if user_id and remember:
+            expires_after = timedelta(days=30).total_seconds()
+        elif user_id:
+            expires_after = timedelta(minutes=60).total_seconds()
+        else:
+            expires_after = timedelta(minutes=10).total_seconds()
+
+        new_session = Session(created_at=isoformat(virtool.utils.timestamp()), ip=ip)
+
+        token = None
+
+        if user_id:
+            token, hashed = virtool.utils.generate_key()
+            new_session.authentication = SessionAuthentication(
+                token=hashed, user_id=user_id
+            )
+
+        await self.redis.set(session_id, dumps(new_session), expire=int(expires_after))
+
+        return session_id, new_session, token
+
+    async def create_session_id(self) -> str:
+        """
+        Create a new unique session id.
+
+        :return: a session id
+
+        """
+        session_id = "session_" + secrets.token_hex(32)
+
+        while await self.redis.get(session_id):
+            session_id = "session_" + secrets.token_hex(32)
+
+        return session_id
+
+    async def get(
+        self, session_id: str, session_token: Optional[str] = None
+    ) -> Tuple[Optional[Session], Optional[str]]:
```

Comment:
The use of tuples in create and replace is not great, but this is the worst. Too many possible paths through the code (cyclomatic complexity) and unclear return values.

I think there is no avoiding splitting up this function based on whether a token is available or not. The `session_token` is doing dual duty as an actual value and as a flag.

I've probably sent you this before: https://alexkondov.com/should-you-pass-boolean-to-functions/.